File: wasm/src/serde/values/duration.rs
use serde::Serialize;
use typst::foundations::Repr;
use tsify::Tsify;

use typst::foundations::Duration as TypstDuration;

#[derive(Serialize, Tsify)]
pub struct DurationValue {
    pub repr: String,
    pub seconds: f64,
    pub minutes: f64,
    pub hours: f64,

    pub days: f64,
    pub weeks: f64,
}

impl From<&TypstDuration> for DurationValue {
    fn from(duration: &TypstDuration) -> Self {
        DurationValue {
            repr: duration.repr().to_string(),
            seconds: duration.seconds(),
            minutes: duration.minutes(),
            hours: duration.hours(),

            days: duration.days(),
            weeks: duration.weeks(),
        }
    }
}

File: wasm/src/serde/values/duration.ts
import type { DurationValue as Duration } from '@wasm';
import type { FormattedValue } from '.';
import { wrapWithTypcInline } from './utils';

export function formatDuration(dur: Duration): FormattedValue {
  const parts: string[] = [];
  if (dur.weeks) parts.push(`${dur.weeks}w`);
  if (dur.days) parts.push(`${dur.days}d`);
  if (dur.hours) parts.push(`${dur.hours}h`);
  if (dur.minutes) parts.push(`${dur.minutes}m`);
  if (dur.seconds) parts.push(`${dur.seconds}s`);

  return { top: wrapWithTypcInline(`${parts.join(' ') || (dur.seconds === 0 ? '0s' : '')}`) };
}
